### ISSUE TITLE
proposal: `avoid_type_ambiguation`

### ISSUE THREAD

**Issue by jacob314** [dart-lang/linter#3463]
## Description
Warn about generic methods that do nothing but take non-nullable values and make them look nullable.

## Details

Due to limitations for how the Dart compiler surfaces warnings, some method have become popular that do nothing other than confuse the compiler enough to not emit diagnostics about unneeded null checks. See https://github.com/dart-lang/sdk/issues/46264#issuecomment-1151476029 for more details.

Now that https://github.com/dart-lang/sdk/issues/46264 is being resolved, these workarounds can and should be removed.
I can't see any purpose of code like this that takes something non-nullable and returns a nullable version so I believe we can safely lint for it.

## Kind

Other: cleanup anti-patterns that have become somewhat popular due to usage in Flutter to work around hacks as part of null safety migrations.

## Good Examples 

```dart
T _ambiguate<T>(T value) => value; // this is fine as it doesn't ambiguate anything.

T? _maybeNull<T>(T? value) => value == foo ? value : null;

T _disambiguate<T>(T? value) => !value;

T get nonNullable => !this;

```
*Add a few examples that demonstrate a “good” adoption of this lint’s principle.*

## Bad Examples

```dart
T? _ambiguate<T>(T? value) => value; // this can quickfix to the harmless:  T _ambiguate<T>(T value) => value;

T? get nullable => this;
```

### Discussion checklist

- [ ] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).  https://github.com/dart-lang/sdk/issues/46264
- [ ] Real world example:  https://github.com/dart-lang/sdk/issues/46264#issuecomment-1151476029

**Comment by bwilkerson** [comment]
While I understand the motivation for this request, I'm not sure the premise is correct (see below). But even assuming the premise is valid, I have to question the long term value of this lint. Once the existing examples are cleaned up, it seems to me to be fairly unlikely that users would create such a method often enough to warrant the lint's existance.

> I can't see any purpose of code like this that takes something non-nullable and returns a nullable version ...
> ...
> Other: cleanup anti-patterns that have become somewhat popular due to usage in Flutter to work around hacks as part of null safety migrations.

See https://github.com/dart-lang/sdk/issues/46264#issuecomment-1156664728. As I understand that comment, the Flutter function was added to allow the team to change an API. It appears to have had nothing to do with null safety migration.

**Comment by stuartmorgan** [comment]
> As I understand that comment, the Flutter function was added to allow the team to change an API. It appears to have had nothing to do with null safety migration.

That's correct; we needed packages to work without warning on Flutter `stable` (where APIs were returning nullable values) and `master` (where the same APIs were returning non-nullable values) at the same time.

**Comment by srawlins** [comment]
Yeah I don't think this occurs enough to be of high value...
